# Seller Metrics

## Cloning my GitHub repo

In [1]:
!git clone https://github.com/kaylatucker/CS690_Capstone.git

Cloning into 'CS690_Capstone'...
remote: Enumerating objects: 152, done.
remote: Counting objects: 100% (112/112), done.
remote: Compressing objects: 100% (105/105), done.
remote: Total 152 (delta 59), reused 13 (delta 7), pack-reused 40 (from 4)
Receiving objects: 100% (152/152), 67.02 MiB | 20.71 MiB/s, done.
Resolving deltas: 100% (61/61), done.


In [2]:
from pathlib import Path
import pandas as pd

REPO_DIR = Path("/content/CS690_Capstone")
PROCESSED_DIR = REPO_DIR / "02_data" / "processed"
RAW_DIR = REPO_DIR / "02_data" / "raw"

## Load the tables needed

In [3]:
items = pd.read_csv(
    PROCESSED_DIR / "order_items_clean.csv",
    parse_dates=["shipping_limit_date"]
)

orders = pd.read_csv(
    PROCESSED_DIR / "orders_clean.csv",
    parse_dates=[
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
)

reviews = pd.read_csv(
    PROCESSED_DIR / "order_reviews_clean.csv",
    parse_dates=[
        "review_creation_date",
        "review_answer_timestamp"
    ]
)

products = pd.read_csv(
    PROCESSED_DIR / "products_clean.csv"
)

sellers = pd.read_csv(
    RAW_DIR / "olist_sellers_dataset.csv",
    dtype={"seller_zip_code_prefix": str}
)

print("Items:", items.shape)
print("Orders:", orders.shape)
print("Reviews:", reviews.shape)
print("Products:", products.shape)
print("Sellers:", sellers.shape)

Items: (112650, 7)
Orders: (99441, 11)
Reviews: (99224, 7)
Products: (32951, 10)
Sellers: (3095, 4)


## Build the basic seller metrics

In [4]:
seller_metrics = (
    items
    .groupby("seller_id", as_index=False)
    .agg(
        seller_order_count=("order_id", "nunique"),
        seller_items_sold=("order_item_id", "count"),
        seller_total_sales=("price", "sum"),
        seller_avg_item_value=("price", "mean"),
        seller_product_count=("product_id", "nunique"),
        seller_total_freight=("freight_value", "sum"),
        seller_avg_freight=("freight_value", "mean")
    )
)

seller_metrics.head()

,seller_id,seller_order_count,seller_items_sold,seller_total_sales,seller_avg_item_value,seller_product_count,seller_total_freight,seller_avg_freight
0,0015a82c2db000af6aaaf3ae2ecb0532,3,3,2685.00,895.000000,1,63.06,21.020000
1,001cca7ae9ae17fb1caed9dfb1094831,200,239,25080.03,104.937364,11,8854.14,37.046611
2,001e6ad469a905060d959994f1b41e4f,1,1,250.00,250.000000,1,17.94,17.940000
3,002100f778ceb8431b7a1020ff7ab48f,51,55,1234.50,22.445455,24,793.66,14.430182
4,003554e2dce176b5555353e4f3555ac8,1,1,120.00,120.000000,1,19.38,19.380000


## Add number of categories sold

In [5]:
items_products = items.merge(
    products[
        ["product_id", "product_category_name_english"]
    ],
    on="product_id",
    how="left",
    validate="many_to_one"
)

seller_categories = (
    items_products
    .groupby("seller_id", as_index=False)
    .agg(
        seller_category_count=(
            "product_category_name_english",
            "nunique"
        )
    )
)

seller_metrics = seller_metrics.merge(
    seller_categories,
    on="seller_id",
    how="left",
    validate="one_to_one"
)

## Prepare reviews correctly

In [6]:
order_reviews = (
    reviews
    .groupby("order_id", as_index=False)
    .agg(
        order_review_score=("review_score", "mean"),
        review_count=("review_score", "count")
    )
)

print("Order-level review rows:", len(order_reviews))
print("Unique order IDs:", order_reviews["order_id"].nunique())

Order-level review rows: 98673
Unique order IDs: 98673


## Identify single-seller orders

In [7]:
order_sellers = (
    items
    .groupby("order_id")
    .agg(
        seller_count=("seller_id", "nunique"),
        seller_id=("seller_id", "first")
    )
    .reset_index()
)

single_seller_orders = order_sellers[
    order_sellers["seller_count"] == 1
].copy()

print("Single-seller orders:", len(single_seller_orders))

Single-seller orders: 97388


## Add customer-experience measures to sellers

In [8]:
seller_order_experience = (
    single_seller_orders
    .merge(
        order_reviews,
        on="order_id",
        how="left",
        validate="one_to_one"
    )
    .merge(
        orders[
            ["order_id", "late_delivery_flag"]
        ],
        on="order_id",
        how="left",
        validate="one_to_one"
    )
)

### aggregate to seller level

In [9]:
seller_experience_metrics = (
    seller_order_experience
    .groupby("seller_id", as_index=False)
    .agg(
        seller_average_review=(
            "order_review_score",
            "mean"
        ),
        seller_reviewed_order_count=(
            "order_review_score",
            "count"
        ),
        seller_late_delivery_rate=(
            "late_delivery_flag",
            "mean"
        ),
        seller_delivery_order_count=(
            "late_delivery_flag",
            "count"
        )
    )
)

seller_metrics = seller_metrics.merge(
    seller_experience_metrics,
    on="seller_id",
    how="left",
    validate="one_to_one"
)

## Add seller location

In [10]:
seller_metrics = seller_metrics.merge(
    sellers,
    on="seller_id",
    how="left",
    validate="one_to_one"
)

## Validate the result

In [11]:
print("Seller metric rows:", len(seller_metrics))
print("Unique seller IDs:", seller_metrics["seller_id"].nunique())
print("Duplicate seller IDs:", seller_metrics["seller_id"].duplicated().sum())

display(seller_metrics.head())

Seller metric rows: 3095
Unique seller IDs: 3095
Duplicate seller IDs: 0


,seller_id,seller_order_count,seller_items_sold,seller_total_sales,seller_avg_item_value,seller_product_count,seller_total_freight,seller_avg_freight,seller_category_count,seller_average_review,seller_reviewed_order_count,seller_late_delivery_rate,seller_delivery_order_count,seller_zip_code_prefix,seller_city,seller_state
0,0015a82c2db000af6aaaf3ae2ecb0532,3,3,2685.00,895.000000,1,63.06,21.020000,1,3.666667,3.0,0.00000,3.0,09080,santo andre,SP
1,001cca7ae9ae17fb1caed9dfb1094831,200,239,25080.03,104.937364,11,8854.14,37.046611,2,4.000000,196.0,0.06701,194.0,29156,cariacica,ES
2,001e6ad469a905060d959994f1b41e4f,1,1,250.00,250.000000,1,17.94,17.940000,1,1.000000,1.0,NaN,0.0,24754,sao goncalo,RJ
3,002100f778ceb8431b7a1020ff7ab48f,51,55,1234.50,22.445455,24,793.66,14.430182,1,3.938776,49.0,0.18750,48.0,14405,franca,SP
4,003554e2dce176b5555353e4f3555ac8,1,1,120.00,120.000000,1,19.38,19.380000,0,5.000000,1.0,0.00000,1.0,74565,goiania,GO


In [12]:
metric_cols = [
    "seller_order_count",
    "seller_items_sold",
    "seller_total_sales",
    "seller_avg_item_value",
    "seller_product_count",
    "seller_category_count",
    "seller_total_freight",
    "seller_avg_freight",
    "seller_average_review",
    "seller_reviewed_order_count",
    "seller_late_delivery_rate",
    "seller_delivery_order_count"
]

display(seller_metrics[metric_cols].describe().T)

,count,mean,std,min,25%,50%,75%,max
seller_order_count,3095.0,32.313409,105.139763,1.00,2.000000,6.000000,21.500000,1854.000000
seller_items_sold,3095.0,36.397415,119.193461,1.00,2.000000,8.000000,24.000000,2033.000000
seller_total_sales,3095.0,4391.484233,13921.997192,3.50,208.850000,821.480000,3280.830000,229472.630000
seller_avg_item_value,3095.0,176.325102,322.145495,3.50,52.183608,95.465385,173.992857,6729.000000
seller_product_count,3095.0,11.130210,24.473252,1.00,2.000000,4.000000,10.000000,399.000000
seller_category_count,3095.0,2.054604,2.098555,0.00,1.000000,1.000000,2.000000,27.000000
seller_total_freight,3095.0,727.595974,2327.157672,6.66,46.170000,155.520000,558.170000,51612.550000
seller_avg_freight,3095.0,23.380116,18.957766,1.20,14.740000,18.230000,24.368333,308.336667
seller_average_review,3073.0,4.036663,0.951055,1.00,3.833333,4.237288,4.666667,5.000000
seller_reviewed_order_count,3078.0,31.401235,102.275869,0.00,2.000000,6.000000,21.000000,1807.000000


## check missingness

In [13]:
display(
    seller_metrics.isna()
    .sum()
    .sort_values(ascending=False)
    .to_frame("missing_count")
)

,missing_count
seller_late_delivery_rate,147
seller_average_review,22
seller_delivery_order_count,17
seller_reviewed_order_count,17
seller_id,0
seller_order_count,0
seller_items_sold,0
seller_total_sales,0
seller_avg_freight,0
seller_total_freight,0


In [14]:
seller_metrics[
    ["seller_reviewed_order_count", "seller_delivery_order_count"]
] = seller_metrics[
    ["seller_reviewed_order_count", "seller_delivery_order_count"]
].fillna(0).astype(int)

## Final validation

In [16]:
# Final validation after count cleanup

print("Seller metric rows:", len(seller_metrics))
print("Unique seller IDs:", seller_metrics["seller_id"].nunique())
print("Duplicate seller IDs:", seller_metrics["seller_id"].duplicated().sum())

print("\nSellers with 0 reviewed orders:",
      (seller_metrics["seller_reviewed_order_count"] == 0).sum())

print("Sellers with 0 valid delivery orders:",
      (seller_metrics["seller_delivery_order_count"] == 0).sum())

display(
    seller_metrics.isna()
    .sum()
    .sort_values(ascending=False)
    .to_frame("missing_count")
)

Seller metric rows: 3095
Unique seller IDs: 3095
Duplicate seller IDs: 0

Sellers with 0 reviewed orders: 22
Sellers with 0 valid delivery orders: 147


,missing_count
seller_late_delivery_rate,147
seller_average_review,22
seller_id,0
seller_order_count,0
seller_avg_item_value,0
seller_product_count,0
seller_items_sold,0
seller_total_sales,0
seller_avg_freight,0
seller_total_freight,0


# Notes

- Built a seller-level dataset with one row per seller (3,095 sellers).
- Seller performance measures include order count, items sold, total sales, average item value, product/category variety, and freight measures.
- Seller sales are based on item prices and do not include freight or payment values.
- Product categories use the cleaned English category field; products with missing category information remain missing and are not counted as unique categories.
- Reviews were first aggregated to one score per order to avoid duplicated order-level review records.
- Review and delivery measures were attributed to sellers only for single-seller orders to avoid ambiguous attribution in multi-seller orders.
- 97,388 orders were eligible for single-seller customer-experience analysis.
- 22 sellers have no eligible reviewed orders, so their average review score remains missing.
- 147 sellers have no valid eligible delivery observations, so their late-delivery rate remains missing.
- Review and delivery observation counts were set to 0 when no eligible observations were available; averages and rates were left missing rather than assigning artificial values.
- Final seller table contains 3,095 unique seller IDs with no duplicate seller rows.

# Export

In [17]:
OUTPUT_FILE = PROCESSED_DIR / "seller_metrics.csv"

seller_metrics.to_csv(
    OUTPUT_FILE,
    index=False
)

print("Saved to:", OUTPUT_FILE)

Saved to: /content/CS690_Capstone/02_data/processed/seller_metrics.csv


### verify file

In [18]:
seller_metrics_check = pd.read_csv(
    OUTPUT_FILE,
    dtype={"seller_zip_code_prefix": str}
)

print("Rows:", len(seller_metrics_check))
print("Unique sellers:", seller_metrics_check["seller_id"].nunique())
print("Duplicate sellers:",
      seller_metrics_check["seller_id"].duplicated().sum())

display(seller_metrics_check.head())

Rows: 3095
Unique sellers: 3095
Duplicate sellers: 0


,seller_id,seller_order_count,seller_items_sold,seller_total_sales,seller_avg_item_value,seller_product_count,seller_total_freight,seller_avg_freight,seller_category_count,seller_average_review,seller_reviewed_order_count,seller_late_delivery_rate,seller_delivery_order_count,seller_zip_code_prefix,seller_city,seller_state
0,0015a82c2db000af6aaaf3ae2ecb0532,3,3,2685.00,895.000000,1,63.06,21.020000,1,3.666667,3,0.00000,3,09080,santo andre,SP
1,001cca7ae9ae17fb1caed9dfb1094831,200,239,25080.03,104.937364,11,8854.14,37.046611,2,4.000000,196,0.06701,194,29156,cariacica,ES
2,001e6ad469a905060d959994f1b41e4f,1,1,250.00,250.000000,1,17.94,17.940000,1,1.000000,1,NaN,0,24754,sao goncalo,RJ
3,002100f778ceb8431b7a1020ff7ab48f,51,55,1234.50,22.445455,24,793.66,14.430182,1,3.938776,49,0.18750,48,14405,franca,SP
4,003554e2dce176b5555353e4f3555ac8,1,1,120.00,120.000000,1,19.38,19.380000,0,5.000000,1,0.00000,1,74565,goiania,GO


### verfying Git recognizes it

In [23]:
!git config --global user.email "ktuck011@odu.edu"
!git config --global user.name "Kayla Tucker"

In [31]:
%cd /content/CS690_Capstone
!git status --short

/content/CS690_Capstone


In [29]:
%cd /content/CS690_Capstone
!cat .gitignore

/content/CS690_Capstone
# Raw and processed data
02_data/raw/*
02_data/processed/*

# Keep folder structure
!02_data/raw/.gitkeep
!02_data/processed/.gitkeep

# Python
__pycache__/
*.pyc
.ipynb_checkpoints/

# Environment files
.env
.venv/
venv/

# OS files
.DS_Store
Thumbs.db


In [32]:
%cd /content/CS690_Capstone
!git branch --show-current

/content/CS690_Capstone
main


In [33]:
!git pull

remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3 (delta 2), reused 0 (delta 0), pack-reused 0 (from 0)
Unpacking objects: 100% (3/3), 1000 bytes | 1000.00 KiB/s, done.
From https://github.com/kaylatucker/CS690_Capstone
   6884942..333cd3b  main       -> origin/main
Updating 6884942..333cd3b
Fast-forward
 .gitignore | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)


In [34]:
!cat .gitignore

# Raw data only
02_data/raw/*

# Keep folder structure
!02_data/raw/.gitkeep

# Python
__pycache__/
*.pyc
.ipynb_checkpoints/

# Environment files
.env
.venv/
venv/

# OS files
.DS_Store
Thumbs.db


In [35]:
!git status --short

?? 02_data/processed/seller_metrics.csv


In [36]:
!git add 02_data/processed/seller_metrics.csv
!git commit -m "Add seller metrics dataset"

[main bd2a341] Add seller metrics dataset
 1 file changed, 3096 insertions(+)
 create mode 100644 02_data/processed/seller_metrics.csv


In [37]:
!git push

fatal: could not read Username for 'https://github.com': No such device or address


In [38]:
from google.colab import userdata
import subprocess

repo = "/content/CS690_Capstone"
token = userdata.get("GITHUB_TOKEN")

branch = subprocess.check_output(
    ["git", "-C", repo, "branch", "--show-current"],
    text=True
).strip()

result = subprocess.run(
    [
        "git",
        "-C", repo,
        "push",
        f"https://x-access-token:{token}@github.com/kaylatucker/CS690_Capstone.git",
        f"HEAD:{branch}"
    ],
    text=True
)

print("Push return code:", result.returncode)

Push return code: 0
